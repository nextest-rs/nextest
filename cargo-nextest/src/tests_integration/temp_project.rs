// Copyright (c) The nextest Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

use camino::{Utf8Path, Utf8PathBuf};
use std::{convert::TryInto, fs, io, path::Path};
use tempfile::TempDir;

// This isn't general purpose -- it specifically excludes certain directories at the root and is
// generally not race-free.
pub(super) fn copy_dir_all(
    src: impl AsRef<Path>,
    dst: impl AsRef<Path>,
    root: bool,
) -> io::Result<()> {
    let src = src.as_ref();
    let dst = dst.as_ref();

    fs::create_dir_all(&dst)?;
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let ty = entry.file_type()?;
        if ty.is_dir() {
            if root && entry.path().file_name() == Some(std::ffi::OsStr::new("target")) {
                continue;
            }
            copy_dir_all(&entry.path(), &dst.join(entry.file_name()), false)?;
        } else {
            fs::copy(entry.path(), dst.join(entry.file_name()))?;
        }
    }
    Ok(())
}

/// A temporary copy of the test project
///
/// This avoid concurrent accesses to the `target` folder.
pub struct TempProject {
    #[allow(dead_code)]
    temp_dir: TempDir,
    workspace_root: Utf8PathBuf,
    target_dir: Utf8PathBuf,
}

impl TempProject {
    pub fn new() -> color_eyre::Result<Self> {
        Self::new_impl(None)
    }

    pub fn new_custom_target_dir(target_dir: &Utf8Path) -> color_eyre::Result<Self> {
        Self::new_impl(Some(target_dir.to_path_buf()))
    }

    fn new_impl(custom_target_dir: Option<Utf8PathBuf>) -> color_eyre::Result<Self> {
        let temp_dir = tempfile::Builder::new()
            .prefix("nextest-fixture")
            .tempdir()?;
        let utf8_path: Utf8PathBuf = temp_dir
            .path()
            .to_path_buf()
            .try_into()
            .expect("tempdir should be valid UTF-8");

        let src_dir = Utf8Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("fixtures/nextest-tests");

        copy_dir_all(&src_dir, &utf8_path, true)?;
        let workspace_root = utf8_path;

        let target_dir = custom_target_dir.unwrap_or_else(|| workspace_root.join("target"));

        Ok(Self {
            temp_dir,
            workspace_root,
            target_dir,
        })
    }

    pub fn workspace_root(&self) -> &Utf8Path {
        &self.workspace_root
    }

    pub fn target_dir(&self) -> &Utf8Path {
        &self.target_dir
    }

    pub fn set_target_dir(&mut self, target_dir: impl Into<Utf8PathBuf>) {
        self.target_dir = target_dir.into();
    }

    pub fn binaries_metadata_path(&self) -> Utf8PathBuf {
        self.target_dir.join("binaries_metadata.json")
    }

    pub fn cargo_metadata_path(&self) -> Utf8PathBuf {
        self.target_dir.join("cargo_metadata.json")
    }

    pub fn manifest_path(&self) -> Utf8PathBuf {
        self.workspace_root.join("Cargo.toml")
    }
}
