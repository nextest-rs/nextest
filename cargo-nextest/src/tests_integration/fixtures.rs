// Copyright (c) The nextest Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

use super::temp_project::TempProject;
use crate::{dispatch::CargoNextestApp, OutputWriter};
use clap::StructOpt;
use nextest_metadata::{
    BinaryListSummary, BuildPlatform, RustTestSuiteStatusSummary, TestListSummary,
};
use once_cell::sync::Lazy;
use regex::Regex;
use std::process::Command;

pub struct TestInfo {
    id: &'static str,
    platform: BuildPlatform,
    test_cases: Vec<(&'static str, bool)>,
}

impl TestInfo {
    fn new(
        id: &'static str,
        platform: BuildPlatform,
        test_cases: Vec<(&'static str, bool)>,
    ) -> Self {
        Self {
            id,
            platform,
            test_cases,
        }
    }
}

pub static EXPECTED_LIST: Lazy<Vec<TestInfo>> = Lazy::new(|| {
    vec![
        TestInfo::new(
            "cdylib-example",
            BuildPlatform::Target,
            vec![("tests::test_multiply_two_cdylib", false)],
        ),
        TestInfo::new(
            "cdylib-link",
            BuildPlatform::Target,
            vec![("test_multiply_two", false)],
        ),
        TestInfo::new("dylib-test", BuildPlatform::Target, vec![]),
        TestInfo::new(
            "nextest-tests::basic",
            BuildPlatform::Target,
            vec![
                ("test_cargo_env_vars", false),
                ("test_cwd", false),
                ("test_execute_bin", false),
                ("test_failure_assert", false),
                ("test_failure_error", false),
                ("test_failure_should_panic", false),
                ("test_flaky_mod_4", false),
                ("test_flaky_mod_6", false),
                ("test_ignored", true),
                ("test_ignored_fail", true),
                ("test_result_failure", false),
                ("test_slow_timeout", true),
                ("test_slow_timeout_2", true),
                ("test_slow_timeout_subprocess", true),
                ("test_stdin_closed", false),
                ("test_subprocess_doesnt_exit", false),
                ("test_success", false),
                ("test_success_should_panic", false),
            ],
        ),
        TestInfo::new(
            "nextest-derive::proc-macro/nextest-derive",
            BuildPlatform::Host,
            vec![("it_works", false)],
        ),
        TestInfo::new(
            "nextest-tests::bench/my-bench",
            BuildPlatform::Target,
            vec![("bench_add_two", false), ("tests::test_execute_bin", false)],
        ),
        TestInfo::new(
            "nextest-tests::bin/nextest-tests",
            BuildPlatform::Target,
            vec![("tests::bin_success", false)],
        ),
        TestInfo::new(
            "nextest-tests",
            BuildPlatform::Target,
            vec![
                ("tests::call_dylib_add_two", false),
                ("tests::unit_test_success", false),
            ],
        ),
        TestInfo::new(
            "nextest-tests::other",
            BuildPlatform::Target,
            vec![("other_test_success", false)],
        ),
        TestInfo::new(
            "nextest-tests::segfault",
            BuildPlatform::Target,
            vec![("test_segfault", false)],
        ),
        TestInfo::new(
            "nextest-tests::bin/other",
            BuildPlatform::Target,
            vec![("tests::other_bin_success", false)],
        ),
        TestInfo::new(
            "nextest-tests::example/nextest-tests",
            BuildPlatform::Target,
            vec![("tests::example_success", false)],
        ),
        TestInfo::new(
            "nextest-tests::example/other",
            BuildPlatform::Target,
            vec![("tests::other_example_success", false)],
        ),
    ]
});

pub fn cargo_bin() -> String {
    match std::env::var("CARGO") {
        Ok(v) => v,
        Err(std::env::VarError::NotPresent) => "cargo".to_owned(),
        Err(err) => panic!("error obtaining CARGO env var: {}", err),
    }
}

#[track_caller]
pub(super) fn set_env_vars() {
    // The dynamic library tests require this flag.
    std::env::set_var("RUSTFLAGS", "-C prefer-dynamic");
    // Set CARGO_TERM_COLOR to never to ensure that ANSI color codes don't interfere with the
    // output.
    // TODO: remove this once programmatic run statuses are supported.
    std::env::set_var("CARGO_TERM_COLOR", "never");
    // This environment variable is required to test the #[bench] fixture. Note that THIS IS FOR
    // TEST CODE ONLY. NEVER USE THIS IN PRODUCTION.
    std::env::set_var("RUSTC_BOOTSTRAP", "1");
}

#[track_caller]
pub fn save_cargo_metadata(p: &TempProject) {
    let mut cmd = Command::new(cargo_bin());
    cmd.args([
        "metadata",
        "--format-version=1",
        "--all-features",
        "--no-deps",
        "--manifest-path",
    ]);
    cmd.arg(p.manifest_path());
    let output = cmd.output().expect("cargo metadata could run");

    assert_eq!(Some(0), output.status.code());
    std::fs::write(p.cargo_metadata_path(), &output.stdout).unwrap();
}

#[track_caller]
pub fn build_tests(p: &TempProject) {
    let args = CargoNextestApp::parse_from([
        "cargo",
        "nextest",
        "--manifest-path",
        p.manifest_path().as_str(),
        "list",
        "--workspace",
        "--all-targets",
        "--message-format",
        "json",
        "--list-type",
        "binaries-only",
        "--target-dir",
        p.target_dir().as_str(),
    ]);

    let mut output = OutputWriter::new_test();
    args.exec(&mut output).unwrap();

    std::fs::write(p.binaries_metadata_path(), output.stdout().unwrap()).unwrap();
}

pub fn check_list_full_output(stdout: &[u8], platform: Option<BuildPlatform>) {
    let result: TestListSummary = serde_json::from_slice(stdout).unwrap();

    let test_suite = &*EXPECTED_LIST;
    assert_eq!(
        test_suite.len(),
        result.rust_suites.len(),
        "test suite counts match"
    );

    for test in test_suite {
        match platform {
            Some(p) if test.platform != p => continue,
            _ => {}
        }

        let entry = result.rust_suites.get(test.id);
        let entry = match entry {
            Some(e) => e,
            _ => panic!("Missing binary: {}", test.id),
        };

        if let Some(platform) = platform {
            if entry.binary.build_platform != platform {
                // The binary should be marked as skipped.
                assert_eq!(
                    entry.status,
                    RustTestSuiteStatusSummary::SKIPPED,
                    "for {}, test suite expected to be skipped because of platform mismatch",
                    test.id
                );
                assert!(
                    entry.test_cases.is_empty(),
                    "skipped test binaries should have no test cases"
                );
                continue;
            }
        }

        assert_eq!(
            entry.status,
            RustTestSuiteStatusSummary::LISTED,
            "for {}, test suite expected to be listed",
            test.id
        );
        assert_eq!(
            test.test_cases.len(),
            entry.test_cases.len(),
            "testcase lengths match for {}",
            test.id
        );
        for case in &test.test_cases {
            let e = entry.test_cases.get(case.0);
            let e = match e {
                Some(e) => e,
                _ => panic!("Missing test case '{}' in '{}'", case.0, test.id),
            };
            assert_eq!(case.1, e.ignored);
        }
    }
}

#[track_caller]
pub fn check_list_binaries_output(stdout: &[u8]) {
    let result: BinaryListSummary = serde_json::from_slice(stdout).unwrap();

    let test_suite = &*EXPECTED_LIST;
    assert_eq!(test_suite.len(), result.rust_binaries.len());

    for test in test_suite {
        let entry = result
            .rust_binaries
            .iter()
            .find(|(_, bin)| bin.binary_id == test.id);
        let entry = match entry {
            Some(e) => e,
            _ => panic!("Missing binary: {}", test.id),
        };

        assert_eq!(test.platform, entry.1.build_platform);
    }
}

fn make_check_result_regex(result: bool, name: &str) -> Regex {
    let name = regex::escape(name);
    if result {
        Regex::new(&format!(r"PASS \[.*\] *{}", name)).unwrap()
    } else {
        Regex::new(&format!(r"(FAIL|ABORT|SIGSEGV) \[.*\] *{}", name)).unwrap()
    }
}

#[track_caller]
pub fn check_run_output(stderr: &[u8], relocated: bool) {
    // This could be made more robust with a machine-readable output,
    // or maybe using quick-junit output

    let output = String::from_utf8(stderr.to_vec()).unwrap();

    println!("{}", output);

    let cwd_pass = !relocated;

    let expected = &[
        (true, "cdylib-link test_multiply_two"),
        (true, "cdylib-example tests::test_multiply_two_cdylib"),
        (true, "nextest-tests::basic test_cargo_env_vars"),
        (true, "nextest-tests::basic test_execute_bin"),
        (true, "nextest-tests::bench/my-bench bench_add_two"),
        (
            true,
            "nextest-tests::bench/my-bench tests::test_execute_bin",
        ),
        (false, "nextest-tests::basic test_failure_error"),
        (false, "nextest-tests::basic test_flaky_mod_4"),
        (true, "nextest-tests::bin/nextest-tests tests::bin_success"),
        (false, "nextest-tests::basic test_failure_should_panic"),
        (true, "nextest-tests::bin/nextest-tests tests::bin_success"),
        (false, "nextest-tests::basic test_failure_should_panic"),
        (true, "nextest-tests::bin/other tests::other_bin_success"),
        (false, "nextest-tests::basic test_result_failure"),
        (true, "nextest-tests::basic test_success_should_panic"),
        (false, "nextest-tests::basic test_failure_assert"),
        (true, "nextest-tests::basic test_stdin_closed"),
        (false, "nextest-tests::basic test_flaky_mod_6"),
        (cwd_pass, "nextest-tests::basic test_cwd"),
        (
            true,
            "nextest-tests::example/nextest-tests tests::example_success",
        ),
        (true, "nextest-tests::other other_test_success"),
        (true, "nextest-tests::basic test_success"),
        (false, "nextest-tests::segfault test_segfault"),
        (true, "nextest-derive::proc-macro/nextest-derive it_works"),
        (
            true,
            "nextest-tests::example/other tests::other_example_success",
        ),
        (true, "nextest-tests tests::unit_test_success"),
    ];

    for (result, name) in expected {
        let reg = make_check_result_regex(*result, name);
        assert!(reg.is_match(&output), "{}: result didn't match", name);
    }

    let summary_reg = if relocated {
        Regex::new(r"Summary \[.*\] *26 tests run: 18 passed \(1 leaky\), 8 failed, 5 skipped")
            .unwrap()
    } else {
        Regex::new(r"Summary \[.*\] *26 tests run: 19 passed \(1 leaky\), 7 failed, 5 skipped")
            .unwrap()
    };
    assert!(
        summary_reg.is_match(&output),
        "summary didn't match (actual output: {}, relocated: {relocated})",
        output
    );
}
