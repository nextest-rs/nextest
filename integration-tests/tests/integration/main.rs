// Copyright (c) The nextest Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Integration tests
//!
//! Running integration tests on Windows is a bit challenging. When using the local cargo-nextest as
//! the test runner:
//!
//! ```
//! cargo run -p cargo-nextest -- nextest run
//! ```
//!
//! This causes:
//! - Execution of `cargo-nextest`
//!     - Execution of `cargo test --no-run`
//!         - Build `cargo-nextest`
//!
//! Which means that cargo will try and replace the cargo-nextest binary that's currently running.
//! This is forbidden on Windows.
//!
//! To solve this issue, we introduce a "cargo-nextest-dup" binary which is exactly the same as
//! cargo-nextest, except it isn't used as the actual test runner. We refer to it with
//! `NEXTEST_BIN_EXE_cargo-nextest-dup`.

use camino::{Utf8Path, Utf8PathBuf};
use nextest_metadata::{BuildPlatform, NextestExitCode};

mod fixtures;
mod temp_project;

use fixtures::*;
use temp_project::TempProject;
use tempfile::TempDir;

#[test]
fn test_list_default() {
    set_env_vars();
    let p = TempProject::new().unwrap();

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "list",
            "--workspace",
            "--all-targets",
            "--message-format",
            "json",
        ])
        .output();

    check_list_full_output(&output.stdout, None);
}

#[test]
fn test_list_full() {
    set_env_vars();
    let p = TempProject::new().unwrap();

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "list",
            "--workspace",
            "--all-targets",
            "--message-format",
            "json",
            "--list-type",
            "full",
        ])
        .output();

    check_list_full_output(&output.stdout, None);
}

#[test]
fn test_list_binaries_only() {
    set_env_vars();
    let p = TempProject::new().unwrap();

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "list",
            "--workspace",
            "--all-targets",
            "--message-format",
            "json",
            "--list-type",
            "binaries-only",
        ])
        .output();

    check_list_binaries_output(&output.stdout);
}

#[test]
fn test_target_dir() {
    set_env_vars();

    let p = TempProject::new().unwrap();

    std::env::set_current_dir(p.workspace_root())
        .expect("changed current directory to workspace root");

    let run_check = |target_dir: &str, extra_args: Vec<&str>| {
        // The test is for the target directory more than for any specific package, so pick a
        // package that builds quickly.
        let output = CargoNextestCli::new()
            .args(["list", "-p", "cdylib-example", "--message-format", "json"])
            .args(extra_args)
            .output();

        let summary = output.decode_test_list_json().unwrap();
        assert_eq!(
            summary.rust_build_meta.target_directory,
            p.workspace_root().join(target_dir),
            "target directory matches"
        );
    };

    // Absolute target directory
    {
        let abs_target_dir = p.workspace_root().join("test-target-dir-abs");
        run_check(
            "test-target-dir-abs",
            vec!["--target-dir", abs_target_dir.as_str()],
        );
    }

    // Relative target directory
    run_check("test-target-dir", vec!["--target-dir", "test-target-dir"]);

    // CARGO_TARGET_DIR env var
    {
        std::env::set_var("CARGO_TARGET_DIR", "test-target-dir-2");
        run_check("test-target-dir-2", vec![]);
    }

    // CARGO_TARGET_DIR env var + --target-dir
    run_check(
        "test-target-dir-3",
        vec!["--target-dir", "test-target-dir-3"],
    );

    cfg_if::cfg_if! {
        if #[cfg(unix)] {
            // Symlink (should not be dereferenced, same as cargo)
            std::fs::create_dir("symlink-target").unwrap();
            std::os::unix::fs::symlink("symlink-target", "symlink-link").unwrap();
            run_check("symlink-link", vec!["--target-dir", "symlink-link"]);
        }
    }
}

#[test]
fn test_list_full_after_build() {
    set_env_vars();

    let p = TempProject::new().unwrap();
    build_tests(&p);

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "list",
            "--binaries-metadata",
            p.binaries_metadata_path().as_str(),
            "--message-format",
            "json",
        ])
        .output();

    check_list_full_output(&output.stdout, None);
}

#[test]
fn test_list_host_after_build() {
    set_env_vars();

    let p = TempProject::new().unwrap();
    build_tests(&p);

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "list",
            "--binaries-metadata",
            p.binaries_metadata_path().as_str(),
            "--message-format",
            "json",
            "-E",
            "platform(host)",
        ])
        .output();

    check_list_full_output(&output.stdout, Some(BuildPlatform::Host));
}

#[test]
fn test_list_target_after_build() {
    set_env_vars();

    let p = TempProject::new().unwrap();
    build_tests(&p);

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "list",
            "--binaries-metadata",
            p.binaries_metadata_path().as_str(),
            "--message-format",
            "json",
            "-E",
            "platform(target)",
        ])
        .output();

    check_list_full_output(&output.stdout, Some(BuildPlatform::Target));
}

#[test]
fn test_run() {
    set_env_vars();

    let p = TempProject::new().unwrap();

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "run",
            "--workspace",
            "--all-targets",
        ])
        .unchecked(true)
        .output();

    assert_eq!(
        output.exit_code,
        Some(NextestExitCode::TEST_RUN_FAILED),
        "correct exit code for command\n{output}"
    );
    check_run_output(&output.stderr, false);
}

#[test]
fn test_run_after_build() {
    set_env_vars();

    let p = TempProject::new().unwrap();
    build_tests(&p);

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "run",
            "--binaries-metadata",
            p.binaries_metadata_path().as_str(),
        ])
        .unchecked(true)
        .output();

    assert_eq!(
        output.exit_code,
        Some(NextestExitCode::TEST_RUN_FAILED),
        "correct exit code for command\n{output}"
    );
    check_run_output(&output.stderr, false);
}

#[test]
fn test_relocated_run() {
    set_env_vars();

    let custom_target_dir = TempDir::new().unwrap();
    let custom_target_path: &Utf8Path = custom_target_dir
        .path()
        .try_into()
        .expect("tempdir is valid UTF-8");
    let p = TempProject::new_custom_target_dir(custom_target_path).unwrap();

    build_tests(&p);
    save_cargo_metadata(&p);

    let mut p2 = TempProject::new().unwrap();
    let new_target_path = p2.workspace_root().join("test-subdir");

    // copy target directory over
    std::fs::create_dir_all(&new_target_path).unwrap();
    temp_project::copy_dir_all(custom_target_path, &new_target_path, false).unwrap();
    // Remove the old target path to ensure that any tests that refer to files within it
    // fail.
    std::fs::remove_dir_all(custom_target_path).unwrap();

    p2.set_target_dir(new_target_path);

    // Use relative paths to the workspace root and target directory to do testing in.
    let current_dir: Utf8PathBuf = std::env::current_dir()
        .expect("able to get current directory")
        .try_into()
        .expect("current directory is valid UTF-8");
    let rel_workspace_root = pathdiff::diff_utf8_paths(p2.workspace_root(), &current_dir)
        .expect("diff of two absolute paths should work");
    let rel_target_dir = pathdiff::diff_utf8_paths(p2.target_dir(), &current_dir)
        .expect("diff of two absolute paths should work");

    // Run relocated tests

    let output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p2.manifest_path().as_str(),
            "run",
            "--binaries-metadata",
            p2.binaries_metadata_path().as_str(),
            "--cargo-metadata",
            p2.cargo_metadata_path().as_str(),
            "--workspace-remap",
            rel_workspace_root.as_str(),
            "--target-dir-remap",
            rel_target_dir.as_str(),
        ])
        .unchecked(true)
        .output();

    assert_eq!(
        output.exit_code,
        Some(NextestExitCode::TEST_RUN_FAILED),
        "correct exit code for command\n{output}"
    );
    check_run_output(&output.stderr, true);
}

#[test]
fn test_run_from_archive() {
    set_env_vars();

    let custom_target_dir = TempDir::new().unwrap();
    let custom_target_path: &Utf8Path = custom_target_dir
        .path()
        .try_into()
        .expect("tempdir is valid UTF-8");
    let p = TempProject::new_custom_target_dir(custom_target_path).unwrap();

    let archive_file = p.temp_root().join("my-archive.tar.zst");

    // Write the archive to the archive_file above.
    _ = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "archive",
            "--archive-file",
            archive_file.as_str(),
            "--workspace",
            "--target-dir",
            p.target_dir().as_str(),
            "--all-targets",
        ])
        .output();

    // Remove the old source and target directories to ensure that any tests that refer to files within
    // it fail.
    std::fs::remove_dir_all(p.workspace_root()).unwrap();
    std::fs::remove_dir_all(p.target_dir()).unwrap();

    let p2 = TempProject::new().unwrap();

    let output = CargoNextestCli::new()
        .args([
            "run",
            "--archive-file",
            archive_file.as_str(),
            "--workspace-remap",
            p2.workspace_root().as_str(),
        ])
        .unchecked(true)
        .output();

    assert_eq!(
        output.exit_code,
        Some(NextestExitCode::TEST_RUN_FAILED),
        "correct exit code for command\n{output}"
    );
    check_run_output(&output.stderr, true);
}

#[test]
fn test_show_config_test_groups() {
    set_env_vars();
    let p = TempProject::new().unwrap();

    let default_profile_output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "show-config",
            "test-groups",
            "--workspace",
            "--all-targets",
        ])
        .output();

    insta::assert_snapshot!(default_profile_output.stdout_as_str());

    let default_profile_all_output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "show-config",
            "test-groups",
            "--workspace",
            "--all-targets",
            "--show-default",
        ])
        .output();

    insta::assert_snapshot!(default_profile_all_output.stdout_as_str());

    let with_retries_output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "show-config",
            "test-groups",
            "--workspace",
            "--all-targets",
            "--profile=with-retries",
        ])
        .output();

    insta::assert_snapshot!(with_retries_output.stdout_as_str());

    let with_retries_all_output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "show-config",
            "test-groups",
            "--workspace",
            "--all-targets",
            "--profile=with-retries",
            "--show-default",
        ])
        .output();

    insta::assert_snapshot!(with_retries_all_output.stdout_as_str());

    let with_termination_output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "show-config",
            "test-groups",
            "--workspace",
            "--all-targets",
            "--profile=with-termination",
        ])
        .output();

    insta::assert_snapshot!(with_termination_output.stdout_as_str());

    let with_termination_all_output = CargoNextestCli::new()
        .args([
            "--manifest-path",
            p.manifest_path().as_str(),
            "show-config",
            "test-groups",
            "--workspace",
            "--all-targets",
            "--profile=with-termination",
            "--show-default",
        ])
        .output();

    insta::assert_snapshot!(with_termination_all_output.stdout_as_str());
}
