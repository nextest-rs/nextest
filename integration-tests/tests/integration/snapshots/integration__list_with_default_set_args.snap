---
source: integration-tests/tests/integration/main.rs
expression: default_set_with_args_output.stdout_as_str()
---
nextest-tests::basic:
    test_stdin_closed
